### ISSUE TITLE
Regression in Shrinking

### ISSUE THREAD

**Issue by cstorey** [BurntSushi/quickcheck#126]
Using the following test case:
```rust
#[cfg(test)]
extern crate quickcheck;

#[cfg(test)]
mod test {
    use quickcheck::quickcheck;
    #[test]
    fn test_can_read_prepared_values() {
        fn test_can_read_prepared_values_qc(vals: Vec<Vec<u8>>) -> bool {
            false
        }
        quickcheck(test_can_read_prepared_values_qc as fn(vals: Vec<Vec<u8>>) -> bool)
    }
}
```

It seems to be that with version `0.2.24`, shrinking works okay, and we get:
```
---- replication_log::test::test_can_read_prepared_values stdout ----
        thread 'replication_log::test::test_can_read_prepared_values' panicked at '[quickcheck] TEST FAILED. Arguments: ([])', /home/ceri/.multirust/toolchains/stable/cargo/registry/src/github.com-88ac128001ac3a9a/quickcheck-0.2.24/src/tester.rs:113
```
When on `0.2.25`, I observe:
```
---- replication_log::test::test_can_read_prepared_values stdout ----
        thread 'replication_log::test::test_can_read_prepared_values' panicked at '[quickcheck] TEST FAILED. Arguments: ([[92, 51, 31, 18, 37, 45, 86, 89, 0, 27, 14, 52, 11, 4
1, 57, 65, 5, 65, 83, 87, 92, 36, 38, 18, 90, 85], ...])', /home/ceri/.multirust/toolchains/stable/cargo/registry/src/github.com-88ac128001ac3a9a/quickcheck-0.2.25/src/tester.rs:116
```

I plan to have a look myself, but It seems worth noting anyway.

**Comment by BurntSushi** [comment]
I'll have a look at this soon, but if someone beats me to it, I might start looking here: https://github.com/BurntSushi/quickcheck/compare/0.2.24...0.2.25#diff-a7670f864b31d4c28b62524be9e1745bR274 --- The specific change there was defining a macro to make adding more `Testable` impls on functions with bigger arity easier. It looks like the code that handles failures is quite a bit different.

There was an overhaul to the `Vec` shrinker, but that appears to have been done before 0.2.24.

**Comment by cstorey** [comment]
That's fair; `git bisect` lead me to commit `f017c01`, which is pretty much as you describe. I've been able to confirm too that the shrinker does still generate the same output on both versions, so that seems likely. 

Okay, so I suspect that the issue is that the new `Testable#result` implementations you mention don't short circuit on their first failure. I suspect that just putting a `return` in the right place would fix that. I'll see about raising a pull request for that.

**Comment by cstorey** [comment]
(it's worthing that #127 is related, or at least will be, once actually filled out)

**Comment by cstorey** [comment]
Thinking a bit more about that, an early return wouldn't give us the smallest possible instance in all cases; it seems likely that this would be sorted as a result of implementing #127.

**Comment by BurntSushi** [comment]
Ideally, we could just take the old code and port it into the macro we have now. The macro might make it tricky, not sure.

**Comment by cstorey** [comment]
Fixed in #129
